src/search/algorithms.clj

```diff
@@ -46,6 +49,59 @@
                new-came-from
                new-cost-so-far)))))
 
+
+
+
+
+
+(defn heuristic-search [children-fn heuristic-fn start-state is-true & {:keys [max-states] :or {max-states 1000000}}]
+  (loop [frontier (pm/priority-map start-state 0)
+         came-from {}
+         visited #{}]
+    (if (or (empty? frontier)
+            (>= (count came-from) max-states)
+            (is-true (first (peek frontier))))
+      [came-from nil (first (peek frontier))]
+      (let [current (first (peek frontier))
+            children (set (children-fn current))
+            unvisited-children (clojure.set/difference children visited)
+            ;Next line is completely useless
+            ;heuristics (map (partial heuristic-fn is-true) unvisited-children)
+            new-frontier (reduce #(assoc %1 %2 (heuristic-fn is-true %2)) (pop frontier) unvisited-children)
+            new-came-from (reduce #(assoc %1 %2 current) came-from unvisited-children)
+            new-visited (clojure.set/union children visited)]
+(recur new-frontier new-came-from new-visited)))))
+
+
+
+(defn A* [children-fn heuristic-fn cost-fn start-state is-true & {:keys [max-states] :or {max-states 1000000}}]
+  (loop [frontier (pm/priority-map start-state 0)
+         came-from {}
+         visited #{}
+         cost-so-far {start-state 0}]
+    (if (or (empty? frontier)
+            (>= (count came-from) max-states)
+            (is-true (first (peek frontier))))
+      [came-from nil (first (peek frontier))]
+      (let [current (first (peek frontier))
+            children (set (children-fn current))
+            current-cost (cost-so-far current)
+            children-costs (reduce #(assoc %1 %2 (+ current-cost (cost-fn current %2))) {} children)
+            children-to-add (filter #(or (not (contains? cost-so-far %))
+                                         (< (children-costs %) (cost-so-far %))) children)
+            unvisited-children (clojure.set/difference children visited)
```

Comment:
You wanted to get rid of `unvisited-children`; the genuinely useful variable is `children-to-add`, and references to `unvisited-children` below should be replaced with `children-to-add`.

You'd never see the problem this mistake causes with your peg puzzle because there's no backtracking, but if you did a problem like N-puzzle with backtracking you'd eventually run into issues.